### PR TITLE
Kimberley Mackenzie- Carets

### PR DESCRIPTION
**Pull request by kimberleyamackenzie**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | For works, I wrote a model method for already_voted? to ensure that users could not vote for the same work twice.|
Describe how you approached testing that model method. What edge cases did you come up with? | I actually struggled with a lot of vote-related testing, and though the functionality works on the application itself, I wasn't able to successfully write a test for it.  The vote-related testing was really difficult for me- specifically assigning works & users to the vote.  The controller consistently was unable to find the assigned ID, regardless of whether I hardcoded into the test or used a fixture.|
Describe an edge case test you wrote for a controller | I tested cases for the Sessions controller to both allow existing users to login, create a new user and login them in simultaneously, and not alow duplicate users to be created.|
What are `session` and `flash`? What is the difference between them? | Flash sends messages from the controller to the views, mostly to track success or failures. Flash messages are one-time-only.  Session keeps track of data throughout a session, or usually until the window is closed or logged out.|
Describe a controller filter you wrote. | For finding a work as part of the works controller. |
What was one thing that you gained more clarity on through this assignment? | A little bit deeper understanding of testing as it applies to Rails and how that is different than our prior work with testing, though I still feel like I didn't know quite enough to effectively test all elements of my application. |
What is the Heroku URL of your deployed application |http://kimberley-media-ranker.herokuapp.com/|
Do you have any recommendations on how we could improve this project for the next cohort? | I think that recreating the HTML & CSS of the given app was a very time-consuming part of this project that didn't feel especially purposeful - I think maybe recreating styling of a given site could be a great exercise as part of HTML/CSS week or with RideShare, and the focus for this project could be more on testing and creating our own styling.|

### CODE REVIEW

**Comment by tildeee** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | x
Comprehension questions	| x
**General** |
Rails fundamentals (RESTful routing, use of named paths) | good job using resources :)
Semantic HTML | x
Errors are reported to the user | x
Business logic lives in the models | x
Models are thoroughly tested, including relations, validations and any custom logic | x
Controllers are thoroughly tested | x, except for votes, which you acknowledge, good job testing otherwise!
**Wave 1 - Media** |
Splash page shows the three media categories | x
Basic CRUD operations on media are present and functional | x
**Wave 2 - Users and Votes** |
Users can log in and log out | x
The ID of the current user is stored in the session | x
Individual user pages and the user list are present | x
A user cannot vote for the same media more than once | x
All media lists are ordered by vote count | x
Splash page contains a media spotlight | not ranked by highest vote
Media pages contain lists of voting users | x
**Wave 3 - Styling** |
Foundation is used appropriately | x
Look and feel is similar to the original | x
**Overall** |

**Comment by tildeee** [comment]
Fantastic job! The app works really well, and hits everything really well.

The testing is great. With regards to testing the Votes controller, you can set up a test with this code:

```
  it "should be able to create a new vote" do
    post login_path, params: {username: users(:test_user).username}
    puts works(:test_book).votes.count
    post work_votes_path(works(:test_book).id)
    puts works(:test_book).votes.count
  end
```

Of course, in that code I just printed some stats instead of asserting on it, but yeah, you should be able to work from there.

Good work overall!

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+class SessionsController < ApplicationController
+
+  def login_form
+  end
+
+  def login
+    user = User.find_by(username: params[:username])
```
Comment (tildeee) [inline app/controllers/sessions_controller.rb:7]: one small bug-- the `username` field in the login_form.html.erb is optional, so you can "log in" with a blank username :) This doesn't really affect the site's experience but yeah!
